web-common/src/features/project/ProjectDeployer.ts

```diff
@@ -0,0 +1,177 @@
+        this.redeployMutation,
+        this.deploying,
+      ],
+      ([validation, deployMutation, redeployMutation, deploying]) => {
+        if (
+          validation.isFetching ||
+          deployMutation.isLoading ||
+          redeployMutation.isLoading ||
+          deploying
+        ) {
+          return {
+            isLoading: true,
+            error: undefined,
+          };
+        }
+
+        return {
+          isLoading: false,
+          error: extractDeployError(
+            (validation.error as ConnectError) ??
+              (deployMutation.error as ConnectError) ??
+              (redeployMutation.error as ConnectError),
+          ).message,
+        };
+      },
+    );
+  }
+
+  public async validate() {
+    let validation = get(this.validation).data as DeployValidationResponse;
+    if (validation?.deployedProjectId) {
+      return true;
+    }
+
+    await waitUntil(() => !get(this.validation).isFetching);
+    validation = get(this.validation).data as DeployValidationResponse;
+
+    if (!validation.isAuthenticated) {
+      void behaviourEvent?.fireDeployEvent(BehaviourEventAction.LoginStart);
+      window.open(
+        `${validation.loginUrl}/?redirect=${get(page).url.toString()}`,
+        "_self",
+      );
+      return false;
+    } else {
+      void behaviourEvent?.fireDeployEvent(BehaviourEventAction.LoginSuccess);
+    }
+
+    // Disabling for now. Will support this though "Connect to github"
+    // if (
+    //   validation.isGithubRepo &&
+    //   (!validation.isGithubConnected || !validation.isGithubRepoAccessGranted)
+    // ) {
+    //   // if the project is a github repo and not connected to github then redirect to grant access
+    //   window.open(`${validation.githubGrantAccessUrl}`, "_self");
+    //   return false;
+    // }
+
+    return true;
+  }
+
+  public async deploy(org?: string) {
+    // safeguard around deploy
+    if (!(await this.validate())) return;
+
+    this.deploying.set(true);
+    try {
+      const validation = get(this.validation).data as DeployValidationResponse;
+      if (validation.deployedProjectId) {
+        const resp = await get(this.redeployMutation).mutateAsync({
+          projectId: validation.deployedProjectId,
+          reupload: !validation.isGithubRepo,
+        });
+        window.open(resp.frontendUrl, "_self");
+      } else {
+        let checkNextOrg = false;
+        if (!org) {
+          if (validation.rillUserOrgs.length === 1) {
+            org = validation.rillUserOrgs[0];
+          } else if (validation.rillUserOrgs.length > 1) {
+            this.promptOrgSelection.set(true);
+            return;
+          } else {
+            const userResp = await queryClient.fetchQuery({
+              queryKey: getLocalServiceGetCurrentUserQueryKey(),
+              queryFn: localServiceGetCurrentUser,
+            });
+            org = userResp.user!.displayName.replace(/ /g, "");
+            checkNextOrg = true;
+          }
```

Comment:
With the `try/catch` and 3 nested `if` statements, this is pretty deep indentation. Consider if there's a way to avoid such nesting, e.g. with guard clauses or factoring out a `getOrg()` function.